In [1]:
#Importing Required Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 

### Read the Dataset

In [2]:
#Reading the dataset
df = pd.read_csv("diabetes_prediction_dataset.csv")
df.head()

gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

In [3]:
df.describe()

age  hypertension  heart_disease            bmi  \
count  100000.000000  100000.00000  100000.000000  100000.000000   
mean       41.885856       0.07485       0.039420      27.320767   
std        22.516840       0.26315       0.194593       6.636783   
min         0.080000       0.00000       0.000000      10.010000   
25%        24.000000       0.00000       0.000000      23.630000   
50%        43.000000       0.00000       0.000000      27.320000   
75%        60.000000       0.00000       0.000000      29.580000   
max        80.000000       1.00000       1.000000      95.690000   

         HbA1c_level  blood_glucose_level       diabetes  
count  100000.000000        100000.000000  100000.000000  
mean        5.527507           138.058060       0.085000  
std         1.070672            40.708136       0.278883  
min         3.500000            80.000000       0.000000  
25%         4.800000           100.000000       0.000000  
50%         5.800000           140.000000       0.000000  
75%         6.200000           159.000000       0.000000  
max         9.000000           300.000000       1.000000

In [4]:
df.columns

Index(['gender', 'age', 'hypertension', 'heart_disease', 'smoking_history',
       'bmi', 'HbA1c_level', 'blood_glucose_level', 'diabetes'],
      dtype='object')

In [5]:
df.head()

gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

In [6]:
df.describe(include="O")


gender smoking_history
count   100000          100000
unique       3               6
top     Female         No Info
freq     58552           35816

## Data Processing

### Null Value Analysis

In [7]:
df.isna()

gender    age  hypertension  heart_disease  smoking_history    bmi  \
0       False  False         False          False            False  False   
1       False  False         False          False            False  False   
2       False  False         False          False            False  False   
3       False  False         False          False            False  False   
4       False  False         False          False            False  False   
...       ...    ...           ...            ...              ...    ...   
99995   False  False         False          False            False  False   
99996   False  False         False          False            False  False   
99997   False  False         False          False            False  False   
99998   False  False         False          False            False  False   
99999   False  False         False          False            False  False   

       HbA1c_level  blood_glucose_level  diabetes  
0            False                False     False  
1            False                False     False  
2            False                False     False  
3            False                False     False  
4            False                False     False  
...            ...                  ...       ...  
99995        False                False     False  
99996        False                False     False  
99997        False                False     False  
99998        False                False     False  
99999        False                False     False  

[100000 rows x 9 columns]

In [8]:
df.isna().sum()

gender                 0
age                    0
hypertension           0
heart_disease          0
smoking_history        0
bmi                    0
HbA1c_level            0
blood_glucose_level    0
diabetes               0
dtype: int64

By obeserving all the columns there are not any null values.

<b>Finding Duplicate Values</b>

In [9]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
99995     True
99996    False
99997    False
99998    False
99999    False
Length: 100000, dtype: bool

In [10]:
df[df.duplicated()]

gender   age  hypertension  heart_disease smoking_history    bmi  \
2756     Male  80.0             0              0         No Info  27.32   
3272   Female  80.0             0              0         No Info  27.32   
3418   Female  19.0             0              0         No Info  27.32   
3939   Female  78.0             1              0          former  27.32   
3960     Male  47.0             0              0         No Info  27.32   
...       ...   ...           ...            ...             ...    ...   
99980  Female  52.0             0              0           never  27.32   
99985    Male  25.0             0              0         No Info  27.32   
99989  Female  26.0             0              0         No Info  27.32   
99990    Male  39.0             0              0         No Info  27.32   
99995  Female  80.0             0              0         No Info  27.32   

       HbA1c_level  blood_glucose_level  diabetes  
2756           6.6                  159         0  
3272           3.5                   80         0  
3418           6.5                  100         0  
3939           3.5                  130         0  
3960           6.0                  200         0  
...            ...                  ...       ...  
99980          6.1                  145         0  
99985          5.8                  145         0  
99989          5.0                  158         0  
99990          6.1                  100         0  
99995          6.2                   90         0  

[3854 rows x 9 columns]

In [11]:
df.duplicated().sum()

3854

We found there are total 3888 duplicate values. Now we will drop all the duplicated values in dataset

In [12]:
df.drop_duplicates(inplace= True)

In [13]:
df[df.duplicated()]

Empty DataFrame
Columns: [gender, age, hypertension, heart_disease, smoking_history, bmi, HbA1c_level, blood_glucose_level, diabetes]
Index: []

Now we have droped all the duplicated values by using drop command. There no duplicated values in the dataset.

In [14]:
#Checking the Info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 96146 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   gender               96146 non-null  object 
 1   age                  96146 non-null  float64
 2   hypertension         96146 non-null  int64  
 3   heart_disease        96146 non-null  int64  
 4   smoking_history      96146 non-null  object 
 5   bmi                  96146 non-null  float64
 6   HbA1c_level          96146 non-null  float64
 7   blood_glucose_level  96146 non-null  int64  
 8   diabetes             96146 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 7.3+ MB


In [15]:
#Correcting the datatypes
df.age = df.age.astype(int)

from above command we change the datatype of age column

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 96146 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   gender               96146 non-null  object 
 1   age                  96146 non-null  int32  
 2   hypertension         96146 non-null  int64  
 3   heart_disease        96146 non-null  int64  
 4   smoking_history      96146 non-null  object 
 5   bmi                  96146 non-null  float64
 6   HbA1c_level          96146 non-null  float64
 7   blood_glucose_level  96146 non-null  int64  
 8   diabetes             96146 non-null  int64  
dtypes: float64(2), int32(1), int64(4), object(2)
memory usage: 7.0+ MB


## Label Encoding 

In [17]:
from sklearn.preprocessing import LabelEncoder
df[["gender", "smoking_history"]] = df[["gender", "smoking_history"]].apply(LabelEncoder().fit_transform)

In [18]:
df.head()

gender  age  hypertension  heart_disease  smoking_history    bmi  \
0       0   80             0              1                4  25.19   
1       0   54             0              0                0  27.32   
2       1   28             0              0                4  27.32   
3       0   36             0              0                1  23.45   
4       1   76             1              1                1  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

## Model Training

In [19]:
X = df.drop("diabetes" , axis = 1)
y = df.diabetes

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=0)

In [21]:
X_train, X_test, y_train, y_test

(       gender  age  hypertension  heart_disease  smoking_history    bmi  \
 72101       1   42             0              0                0  30.13   
 9504        1   62             0              1                3  40.90   
 39541       0   62             0              0                1  41.52   
 67858       1    4             0              0                0  27.32   
 11335       1   65             0              0                4  28.25   
 ...       ...  ...           ...            ...              ...    ...   
 21442       1   48             0              0                0  34.87   
 46806       0   73             0              1                3  31.74   
 43413       0   66             0              0                3  41.00   
 44393       1    6             0              0                0  15.02   
 70291       1   38             0              0                4  27.32   
 
        HbA1c_level  blood_glucose_level  
 72101          6.0                   80  


In [22]:
from sklearn.metrics import f1_score, accuracy_score, mean_squared_error, classification_report, confusion_matrix

## Decision Tree Classifier

In [23]:
from sklearn.tree import DecisionTreeClassifier

dt_classifier = DecisionTreeClassifier()
dt_classifier.fit(X_train, y_train)

DecisionTreeClassifier()

In [24]:
y_pred_dt= dt_classifier.predict(X_test)
y_pred_dt

array([0, 1, 0, ..., 0, 0, 0], dtype=int64)

In [25]:
print("Accuracy Score: ", accuracy_score(y_test, y_pred_dt) * 100)

Accuracy Score:  94.98179927197089


In [26]:
print("Mean Squared Error: ", mean_squared_error(y_test, y_pred_dt))

Mean Squared Error:  0.050182007280291215


In [27]:
print("F1 Score: ", f1_score(y_test, y_pred_dt))

F1 Score:  0.7180835524393806


In [28]:
print("Classification Report\n", classification_report(y_test, y_pred_dt))

Classification Report
               precision    recall  f1-score   support

           0       0.97      0.97      0.97     17562
           1       0.70      0.74      0.72      1668

    accuracy                           0.95     19230
   macro avg       0.84      0.85      0.85     19230
weighted avg       0.95      0.95      0.95     19230



In [29]:
print("Confusion Matris\n", confusion_matrix(y_test, y_pred_dt))

Confusion Matris
 [[17036   526]
 [  439  1229]]


## Random Forest Classifier

In [30]:
from sklearn.ensemble import RandomForestClassifier
rf_classifier = RandomForestClassifier(n_estimators=100)
rf_classifier.fit(X_train, y_train)

RandomForestClassifier()

In [31]:
y_pred_rf = rf_classifier.predict(X_test)
y_pred_rf

array([0, 1, 0, ..., 0, 0, 0], dtype=int64)

In [32]:
print("Accuracy Score: ", accuracy_score(y_test, y_pred_rf) * 100)

Accuracy Score:  96.9578783151326


In [33]:
print("Mean Squared Error: ", mean_squared_error(y_test, y_pred_rf))

Mean Squared Error:  0.030421216848673948


In [34]:
print("F1 Score: ", f1_score(y_test, y_pred_rf))

F1 Score:  0.7990381312263827


In [35]:
print("Classification Report\n", classification_report(y_test, y_pred_rf))

Classification Report
               precision    recall  f1-score   support

           0       0.97      1.00      0.98     17562
           1       0.94      0.70      0.80      1668

    accuracy                           0.97     19230
   macro avg       0.95      0.85      0.89     19230
weighted avg       0.97      0.97      0.97     19230



In [44]:
print("Confusion Matrics\n", confusion_matrix(y_test, y_pred_rf))

Confusion Matrics
 [[17482    80]
 [  505  1163]]


## XG Boost


In [37]:
from xgboost import XGBClassifier
xgb_classifier = XGBClassifier(learning_rate = 0.5, n_estimators = 100)
xgb_classifier.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.5, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [38]:
y_pred_xgb = xgb_classifier.predict(X_test)
y_pred_xgb

array([0, 1, 0, ..., 0, 0, 0])

In [39]:
print("Accuracy Score: ", accuracy_score(y_test, y_pred_xgb) * 100)

Accuracy Score:  96.98387935517421


In [40]:
print("Mean Squared Error: ", mean_squared_error(y_test, y_pred_xgb))

Mean Squared Error:  0.03016120644825793


In [41]:
print("F1 Score: ", f1_score(y_test, y_pred_xgb))

F1 Score:  0.8029891304347826


In [42]:
print("Classification Report\n", classification_report(y_test, y_pred_xgb))

Classification Report
               precision    recall  f1-score   support

           0       0.97      0.99      0.98     17562
           1       0.93      0.71      0.80      1668

    accuracy                           0.97     19230
   macro avg       0.95      0.85      0.89     19230
weighted avg       0.97      0.97      0.97     19230



In [43]:
print("Confusion Matris\n", confusion_matrix(y_test, y_pred_xgb))

Confusion Matris
 [[17468    94]
 [  486  1182]]


### Conclusion

* <b>After analyzing and comparing the performance of three machine learning algorithms, namely RandomForest, XGBoost, and decision tree, in predicting diabetes, we can draw the following conclusions:</b>


* <b>Firstly, all three models showed promising results with high accuracy scores. However, the XGBoost model out performed the highest accuracy score of 96.98 % on the test data set. </b>


* <b>Secondly, the Random Forest and decision tree models also showed good performance with accuracy scores of 96.95 % and 94.98 % respectively. Additionally, both models also provided insights into the most important features contributing to the prediction of diabetes, which can be useful for future research and clinical practice. </b>


* <b>Overall, machine learning algorithms have the potential to aid in the diagnosis and prediction of diabetes, with XGBoost being the most accurate among the three models tested. However, it is important to note that these models should not be used as a replacement for clinical judgment, and further research is needed to evaluate their performance in larger and more diverse datasets.</b>